src/core/events/handler.py

```diff
@@ -70,53 +70,50 @@ def subscribe_event(self, event, hook=None, predicate=None):
             logging.debug('{} subscribed to {}'.format(hook, event))
 
     def apply_filter(self, event):
-        # filters a given event, if the event should be removed, returns True, 
-        status = None
+        new_event = event
         for hooked_event in self.filters.keys():
             if hooked_event in event.__class__.__mro__:
                 for filter_hook, predicate in self.filters[hooked_event]:
                     if predicate and not predicate(event):
                         continue
 
                     logging.debug('Event {} got filtered with {}'.format(event.__class__, filter_hook))
-                    status = filter_hook(event).execute()
-        return status
+                    new_event = filter_hook(new_event).execute()
+                    # If event is filtered out, stop
+                    if not new_event:
+                        return None
 
-    # getting instantiated event object
-    def publish_event(self, event, caller=None):
-        remove_event = False
-        apply_filter_original_code = self.apply_filter.__func__.__code__
+        return new_event
 
+     # getting instantiated event object
+    def publish_event(self, event, caller=None):
         # setting event chain
         if caller:
             event.previous = caller.event
             event.hunter = caller.__class__
 
+        filter_executed = False
         # publishing to subscribers
         for hooked_event in self.hooks.keys():
-            if remove_event:
-                break
             if hooked_event in event.__class__.__mro__:
                 for hook, predicate in self.hooks[hooked_event]:
                     if predicate and not predicate(event):
                         continue
 
-                    # will run once for each event
-                    if self.apply_filter(event):
-                        remove_event = True
-                        break
-                    # formatting the apply_filter method, to not run next time again on the same event.
-                    self.apply_filter.__func__.__code__ = (lambda self, event: None).__code__
-   
+                    # We want to run filters only once and only if we gonna publish it
```

Comment:
Why not keep this simply and just do `apply_filter` on event outside all of this loop, right at the start of this `publish_event` function?

Reply:
The filtering loop is time consuming, by doing it only when the event gets published, we minimize run time. it's for efficiency purposes.

Reply:
I see your point. But:
1. If we will have more 'active' hunters the ratio of none-relevant events may change
2. The current implementation (without python internals) is much easier to understand
3. This is 'core' function - that is only few will be changing it (or read for that matter). Little complexity on account of readability is allowed here I think.

So it can go either way. I will leave it to you to decide.